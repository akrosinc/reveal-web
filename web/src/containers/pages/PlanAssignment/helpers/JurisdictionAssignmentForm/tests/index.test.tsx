import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import MockDate from 'mockdate';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { raZambiaNode } from '../../../../../../components/TreeWalker/tests/fixtures';
import { plans } from '../../../../../../store/ducks/opensrp/PlanDefinition/tests/fixtures';
import { JurisdictionAssignmentForm } from '../index';
import { apiCall, submitCallbackPayload } from './fixtures';

/* tslint:disable-next-line no-var-requires */
const fetch = require('jest-fetch-mock');
jest.mock('../../../../../../configs/env');

describe('PlanAssignment/JurisdictionAssignmentForm', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    MockDate.set('12/30/2019', 0);
  });

  it('renders without crashing', () => {
    if (!raZambiaNode) {
      fail();
    }

    const props = {
      defaultValue: [{ label: 'Team X', value: 'x' }],
      jurisdiction: raZambiaNode,
      options: [
        { label: 'Team X', value: 'x' },
        { label: 'Team Y', value: 'y' },
      ],
      plan: plans[1],
    };

    const wrapper = mount(<JurisdictionAssignmentForm {...props} />);

    expect(wrapper.find('Form').length).toEqual(1);
    expect(wrapper.find('SelectField').props() as any).toMatchSnapshot({
      form: expect.any(Object),
    });
    expect(toJson(wrapper.find('button'))).toMatchSnapshot('Submit buttons');
    wrapper.unmount();
  });

  it('shows error if no plan or jurisdiction', () => {
    const props = {
      defaultValue: [{ label: 'Team X', value: 'x' }],
      options: [
        { label: 'Team X', value: 'x' },
        { label: 'Team Y', value: 'y' },
      ],
    };

    const wrapper = mount(<JurisdictionAssignmentForm {...props} />);

    expect(toJson(wrapper.find('span'))).toMatchSnapshot('no plan or jurisdiction');
    wrapper.unmount();
  });

  it('form works as expected', async () => {
    if (!raZambiaNode) {
      fail();
    }

    fetch.mockResponseOnce(JSON.stringify({}), { status: 200 });

    const cancelMock: any = jest.fn();
    const submitMock: any = jest.fn();
    const growlMock: any = jest.fn();

    const props = {
      cancelCallBackFunc: cancelMock,
      defaultValue: [{ label: 'Team X', value: 'x' }],
      jurisdiction: raZambiaNode,
      options: [
        { label: 'Team X', value: 'x' },
        { label: 'Team Y', value: 'y' },
      ],
      plan: plans[1],
      submitCallBackFunc: submitMock,
      successNotifierBackFunc: growlMock,
    };

    const wrapper = mount(<JurisdictionAssignmentForm {...props} />);

    await act(async () => {
      wrapper.find('button.cancel').simulate('click');
    });
    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(cancelMock).toHaveBeenCalledTimes(1);

    await act(async () => {
      wrapper.find('form').simulate('submit');
    });
    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(fetch.mock.calls).toEqual(apiCall);
    expect(submitMock).toHaveBeenCalledTimes(1);
    expect(submitMock.mock.calls).toEqual(submitCallbackPayload);
    expect(growlMock.mock.calls).toEqual([['Team(s) assignment updated successfully']]);

    wrapper.unmount();
  });

  it('form errors as expected', async () => {
    if (!raZambiaNode) {
      fail();
    }

    fetch.mockResponseOnce(JSON.stringify({}), { status: 400 });

    const submitMock: any = jest.fn();

    const props = {
      defaultValue: [{ label: 'Team X', value: 'x' }],
      jurisdiction: raZambiaNode,
      options: [
        { label: 'Team X', value: 'x' },
        { label: 'Team Y', value: 'y' },
      ],
      plan: plans[1],
      submitCallBackFunc: submitMock,
    };

    const wrapper = mount(<JurisdictionAssignmentForm {...props} />);

    await act(async () => {
      wrapper.find('form').simulate('submit');
    });
    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(submitMock).toHaveBeenCalledTimes(0);
    expect(toJson(wrapper.find('p.assignments-error'))).toMatchSnapshot('Form error');

    wrapper.unmount();
  });
});
