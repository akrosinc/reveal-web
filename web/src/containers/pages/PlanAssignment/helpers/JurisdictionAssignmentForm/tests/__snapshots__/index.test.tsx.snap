// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanAssignment/JurisdictionAssignmentForm form errors as expected: Form error 1`] = `
<p
  className="form-text text-danger assignments-error"
>
  OpenSRPService create on organization/assignLocationsAndPlans failed, HTTP status 400
</p>
`;

exports[`PlanAssignment/JurisdictionAssignmentForm renders without crashing 1`] = `
Object {
  "children": undefined,
  "className": "form-control",
  "defaultValue": Array [
    Object {
      "label": "Team X",
      "value": "x",
    },
  ],
  "field": Object {
    "name": "organizations",
    "onBlur": [Function],
    "onChange": [Function],
    "value": Array [
      "x",
    ],
  },
  "form": Any<Object>,
  "options": Array [
    Object {
      "label": "Team X",
      "value": "x",
    },
    Object {
      "label": "Team Y",
      "value": "y",
    },
  ],
}
`;

exports[`PlanAssignment/JurisdictionAssignmentForm renders without crashing: Submit buttons 1`] = `
Array [
  <button
    aria-label={null}
    className="btn-block cancel btn btn-secondary btn-sm"
    onClick={[Function]}
    type="button"
  >
    Close
  </button>,
  <button
    aria-label={null}
    className="btn-block submit btn btn-primary btn-sm"
    disabled={false}
    onClick={[Function]}
    type="submit"
  >
    Save
  </button>,
]
`;

exports[`PlanAssignment/JurisdictionAssignmentForm shows error if no plan or jurisdiction: no plan or jurisdiction 1`] = `
<span>
  Could not load the form.
</span>
`;
